src/protocol/context/mod.rs

```diff
@@ -0,0 +1,94 @@
+use crate::ff::Field;
+use crate::helpers::messaging::{Gateway, Mesh};
+use crate::helpers::Role;
+use crate::protocol::mul::SecureMul;
+use crate::protocol::prss::{
+    Endpoint as PrssEndpoint, IndexedSharedRandomness, SequentialSharedRandomness,
+};
+use crate::protocol::reveal::Reveal;
+use crate::protocol::{Step, Substep};
+use crate::secret_sharing::SecretSharing;
+use std::sync::Arc;
+
+mod malicious;
+mod semi_honest;
+
+pub use malicious::MaliciousProtocolContext;
+pub use semi_honest::SemiHonestProtocolContext;
+
+/// Context used by each helper to perform secure computation. Provides access to shared randomness
+/// generator and communication channel.
+pub trait ProtocolContext<F: Field>:
+    Clone + SecureMul<F, Share = <Self as ProtocolContext<F>>::Share> + Reveal
+{
+    /// Secret sharing type this context supports.
+    type Share: SecretSharing<F>;
+
+    /// The role of this context.
+    fn role(&self) -> Role;
+
+    /// A unique identifier for this stage of the protocol execution.
+    #[must_use]
+    fn step(&self) -> &Step;
+
+    /// Make a sub-context.
+    /// Note that each invocation of this should use a unique value of `step`.
+    #[must_use]
+    fn narrow<S: Substep + ?Sized>(&self, step: &S) -> Self;
+
+    /// Get the indexed PRSS instance for this step.  It is safe to call this function
+    /// multiple times.
+    ///
+    /// # Panics
+    /// If `prss_rng()` is invoked for the same context, this will panic.  Use of
+    /// these two functions are mutually exclusive.
+    #[must_use]
+    fn prss(&self) -> Arc<IndexedSharedRandomness>;
+
+    /// Get a pair of PRSS-based RNGs.  The first is shared with the helper to the "left",
+    /// the second is shared with the helper to the "right".
+    ///
+    /// # Panics
+    /// This method can only be called once.  This is also mutually exclusive with `prss()`.
+    /// This will panic if you have previously invoked `prss()`.
+    #[must_use]
+    fn prss_rng(&self) -> (SequentialSharedRandomness, SequentialSharedRandomness);
+
+    /// Get a set of communications channels to different peers.
+    #[must_use]
+    fn mesh(&self) -> Mesh<'_, '_>;
+
+    /// Generates a new share of one
+    fn share_of_one(&self) -> <Self as ProtocolContext<F>>::Share;
+}
+
+/// Contains things that are applicable to any implementation of protocol context as see it today
+/// Every context requires access to current step, PRSS and communication and that is what this
+/// struct carries.
+#[derive(Clone, Debug)]
+struct ContextInner<'a> {
+    role: Role,
+    step: Step,
```

Comment:
My thinking (see #139) had `Step` outside of the shared part.  That is, the shared components are kept in the inner piece, but the part that needs to change is kept with each instance.

As long as the inner object is `Sync`, then we should be able to refer to it with `Arc`.  `Arc` is just a pointer, so cloning it is as cheap as copying that and doing a relaxed atomic increment.

Reply:
I can take a stab at moving step away from inner.

Reply:
all right, I've opened a can of worms here.  I tried to avoid `Arc` but I don't seem that I can do it. The most annoying thing is that we really need `Arc` for tests only. `Arc` causes double-indirection every time we need to dereference `Gateway` or `Participant` :(  

Is using `Arc` ok for now or is there a better way to handle this? Also I think if we migrate all tests to use #261, we don't have this problem so should we push these first? 

```mermaid
graph TD;
    TestWorld-->|owns|Gateway;
    ContextInner-->|borrows|Gateway;
    subgraph ephemeral
    Context-->|borrows|ContextInner;
    end
    Test-->|owns|Context;
    Test-->|owns|TestWorld;
```

Reply:
it is done and both contexts now use `Arc`. I've wasted another couple of hours trying to make `MaliciousContext` look like `SemiHonest` or, more precisely `Context<F, Share = Replicated<F>>`.  So sad it didn't work (https://github.com/rust-lang/rust/issues/20400)